playground/models/networks/sampling.py

```diff
@@ -1,4 +1,20 @@
-from keras.layers import Dense, Concatenate, Add
+from keras.layers import Dense, Concatenate, Add, Lambda, Reshape
+from keras import backend as K
+from numpy import newaxis
+
+
+def outer_product(inputs):
+    """
+    inputs: list of two tensors (of equal dimensions,
+        for which you need to compute the outer product
+    """
+    x, y = inputs
+    feature_size = x._shape_as_list()[1] ** 2
```

Comment:
you are accessing a private attribute `_shape_as_list`, which you technically can in Python but from a software engineering point of you should not. You can get the sample dimension in a more legal way using `shape` from the `keras.backend` module, as I have done it below in the sampling functions.